### PR TITLE
Topic: PyTorch DataLoader implementation

### PR DESCRIPTION
**Pull request by YrrepNoj**
Description:
  * Added a utility function that initializes an IWPDatset and returns a
    PyTorch DataLoader that iterates through BATCH_SIZE subsets of the
    dataset.

### CODE REVIEW

File: python/iwp/data_loader.py
Comment (gthomsen) [inline python/iwp/data_loader.py]: Unless there is some exceptional need, this should be imported in alphabetical order, so after copy and before torch.utils.data.dataset.

File: python/iwp/data_loader.py
Comment (gthomsen) [inline python/iwp/data_loader.py]: You should review your branches for extraneous changes tagging along for the ride before pushing.  Adding/removing whitespace, modifying comment ("the" -> "te" below), etc.

File: python/iwp/data_loader.py
Comment (gthomsen) [inline python/iwp/data_loader.py]: This debug print() needs to be removed.

File: python/iwp/data_loader.py
Comment (gthomsen) [inline python/iwp/data_loader.py]: Nit: The column this is wrapped at is inconsistent with the argument description below.  I'd aim for ~80 columns (not hard until 100+) of what is printed via help() rather than in the code itself - that is, the actual line can run ~85 columns since it is in an indented docstring.

File: python/iwp/data_loader.py
Comment (gthomsen) [inline python/iwp/data_loader.py]: We'll want a smaller default, say in the 32-128 range, until we have more experience with the model architecture.  The architecture itself will drive what the batch size is, particularly the largest batch size target GPUs can hold in memory while training.

File: python/iwp/data_loader.py
Comment (gthomsen) [inline python/iwp/data_loader.py]: The default specified here is inconsistent with the actual default.  What does num_workers=1 do?  Spawn one background process independent of the main?

Style: Use number_workers instead of num_workers.

File: python/iwp/data_loader.py
Comment (gthomsen) [inline python/iwp/data_loader.py]: Style: Use shuffle_flag instead of shuffle to be consistent with the code base's naming convention.

File: python/iwp/data_loader.py
Comment (gthomsen) [inline python/iwp/data_loader.py]: Why not provide a class method that acts as a DataLoader factory?  You take a superset of arguments to initialize a Dataset and have three arguments (batch_size, shuffle_flag, number_workers) to create the DataLoader.

It may be poor design to create a DataLoader from a Dataset, though you're better positioned to understand the separation between those two and which approach is more natural.  I'd recommend looking at some PyTorch models (pick a model zoo) and see what patterns exist, if any, to help make this decision.

File: python/iwp/data_loader.py
Comment (gthomsen) [inline python/iwp/data_loader.py]: Nit: Missing a period for the complete sentence.  Also non-docstring comments aren't capitalized, so this isn't consistent with the rest of the code base.

File: python/iwp/data_loader.py
Comment (gthomsen) [inline python/iwp/data_loader.py:626]: Nit: Remove one of the newlines here.

File: python/iwp/data_loader.py
Comment (gthomsen) [inline python/iwp/data_loader.py]: Nit: XY instead of xy for consistency.

**Review by gthomsen — CHANGES_REQUESTED** [review]
You need some research how PyTorch models separate their Datasets and DataLoaders, as we well as address the style issues before we merge.

File: python/iwp/data_loader.py
Comment (YrrepNoj) [reply python/iwp/data_loader.py]: I'm honestly not 100% certain but it seems like that is the case. 

https://discuss.pytorch.org/t/guidelines-for-assigning-num-workers-to-dataloader/813/2

**Comment by gthomsen** [comment]
This has become OBE.  Closing.

**Pull request closed without merging**
